File: frontend/src/api/campaignBuilder.ts
/**
 * Stratum AI - Campaign Builder API Hooks
 *
 * React Query hooks for the Campaign Builder feature:
 * - Platform connectors
 * - Ad accounts
 * - Campaign drafts
 * - Publish logs
 */

import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiClient } from './client'

// =============================================================================
// Types
// =============================================================================

export type Platform = 'meta' | 'google' | 'tiktok' | 'snapchat'
export type ConnectionStatus = 'connected' | 'expired' | 'error' | 'disconnected'
export type DraftStatus = 'draft' | 'submitted' | 'approved' | 'rejected' | 'publishing' | 'published' | 'failed'
export type PublishResult = 'success' | 'failure'

export interface ConnectorStatus {
  platform: string
  status: ConnectionStatus
  connected_at?: string
  last_refreshed_at?: string
  scopes: string[]
  last_error?: string
}

export interface AdAccount {
  id: string
  platform: string
  platform_account_id: string
  name: string
  business_name?: string
  currency: string
  timezone: string
  is_enabled: boolean
  daily_budget_cap?: number
  last_synced_at?: string
}

export interface CampaignDraft {
  id: string
  platform: string
  ad_account_id?: string
  name: string
  description?: string
  status: DraftStatus
  draft_json: Record<string, any>
  created_at: string
  updated_at: string
  submitted_at?: string
  approved_at?: string
  rejected_at?: string
  rejection_reason?: string
  platform_campaign_id?: string
  published_at?: string
}

export interface PublishLog {
  id: string
  draft_id?: string
  platform: string
  platform_account_id: string
  event_time: string
  result_status: PublishResult
  platform_campaign_id?: string
  error_code?: string
  error_message?: string
  retry_count: number
}

export interface CreateDraftPayload {
  platform: string
  ad_account_id: string
  name: string
  description?: string
  draft_json: Record<string, any>
}

export interface UpdateDraftPayload {
  name?: string
  description?: string
  draft_json?: Record<string, any>
}

// =============================================================================
// Connector Hooks
// =============================================================================

export function useConnectorStatus(tenantId: number, platform: Platform) {
  return useQuery({
    queryKey: ['connector-status', tenantId, platform],
    queryFn: async () => {
      const response = await apiClient.get<{ data: ConnectorStatus }>(
        `/campaign-builder/tenant/${tenantId}/connect/${platform}/status`
      )
      return response.data.data
    },
    enabled: !!tenantId && !!platform,
  })
}

export function useStartConnection(tenantId: number) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: async (platform: Platform) => {
      const response = await apiClient.post<{ data: { oauth_url: string } }>(
        `/campaign-builder/tenant/${tenantId}/connect/${platform}/start`
      )
      return response.data.data
    },
    onSuccess: (_, platform) => {
      queryClient.invalidateQueries({ queryKey: ['connector-status', tenantId, platform] })
    },
  })
}

export function useRefreshToken(tenantId: number) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: async (platform: Platform) => {
      const response = await apiClient.post(`/campaign-builder/tenant/${tenantId}/connect/${platform}/refresh`)
      return response.data
    },
    onSuccess: (_, platform) => {
      queryClient.invalidateQueries({ queryKey: ['connector-status', tenantId, platform] })
    },
  })
}

export function useDisconnectPlatform(tenantId: number) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: async (platform: Platform) => {
      const response = await apiClient.delete(`/campaign-builder/tenant/${tenantId}/connect/${platform}`)
      return response.data
    },
    onSuccess: (_, platform) => {
      queryClient.invalidateQueries({ queryKey: ['connector-status', tenantId, platform] })
      queryClient.invalidateQueries({ queryKey: ['ad-accounts', tenantId, platform] })
    },
  })
}

// =============================================================================
// Ad Account Hooks
// =============================================================================

export function useAdAccounts(tenantId: number, platform: Platform, enabledOnly = false) {
  return useQuery({
    queryKey: ['ad-accounts', tenantId, platform, enabledOnly],
    queryFn: async () => {
      const response = await apiClient.get<{ data: AdAccount[] }>(
        `/campaign-builder/tenant/${tenantId}/ad-accounts/${platform}`,
        { params: { enabled_only: enabledOnly } }
      )
      return response.data.data
    },
    enabled: !!tenantId && !!platform,
  })
}

export function useSyncAdAccounts(tenantId: number) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: async (platform: Platform) => {
      const response = await apiClient.post(`/campaign-builder/tenant/${tenantId}/ad-accounts/${platform}/sync`)
      return response.data
    },
    onSuccess: (_, platform) => {
      // Delay refetch to allow sync to complete
      setTimeout(() => {
        queryClient.invalidateQueries({ queryKey: ['ad-accounts', tenantId, platform] })
      }, 2000)
    },
  })
}

export function useUpdateAdAccount(tenantId: number) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: async ({
      platform,
      accountId,
      data,
    }: {
      platform: Platform
      accountId: string
      data: { is_enabled?: boolean; daily_budget_cap?: number }
    }) => {
      const response = await apiClient.put<{ data: AdAccount }>(
        `/campaign-builder/tenant/${tenantId}/ad-accounts/${platform}/${accountId}`,
        data
      )
      return response.data.data
    },
    onSuccess: (_, { platform }) => {
      queryClient.invalidateQueries({ queryKey: ['ad-accounts', tenantId, platform] })
    },
  })
}

// =============================================================================
// Campaign Draft Hooks
// =============================================================================

export function useCampaignDrafts(
  tenantId: number,
  filters?: {
    platform?: Platform
    status?: DraftStatus
    ad_account_id?: string
  }
) {
  return useQuery({
    queryKey: ['campaign-drafts', tenantId, filters],
    queryFn: async () => {
      const response = await apiClient.get<{ data: CampaignDraft[] }>(
        `/campaign-builder/tenant/${tenantId}/campaign-drafts`,
        { params: filters }
      )
      return response.data.data
    },
    enabled: !!tenantId,
  })
}

export function useCampaignDraft(tenantId: number, draftId: string) {
  return useQuery({
    queryKey: ['campaign-draft', tenantId, draftId],
    queryFn: async () => {
      const response = await apiClient.get<{ data: CampaignDraft }>(
        `/campaign-builder/tenant/${tenantId}/campaign-drafts/${draftId}`
      )
      return response.data.data
    },
    enabled: !!tenantId && !!draftId,
  })
}

export function useCreateCampaignDraft(tenantId: number) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: async (data: CreateDraftPayload) => {
      const response = await apiClient.post<{ data: CampaignDraft }>(
        `/campaign-builder/tenant/${tenantId}/campaign-drafts`,
        data
      )
      return response.data.data
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['campaign-drafts', tenantId] })
    },
  })
}

export function useUpdateCampaignDraft(tenantId: number, draftId: string) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: async (data: UpdateDraftPayload) => {
      const response = await apiClient.put<{ data: CampaignDraft }>(
        `/campaign-builder/tenant/${tenantId}/campaign-drafts/${draftId}`,
        data
      )
      return response.data.data
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['campaign-drafts', tenantId] })
      queryClient.invalidateQueries({ queryKey: ['campaign-draft', tenantId, draftId] })
    },
  })
}

export function useSubmitDraft(tenantId: number) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: async (draftId: string) => {
      const response = await apiClient.post<{ data: CampaignDraft }>(
        `/campaign-builder/tenant/${tenantId}/campaign-drafts/${draftId}/submit`
      )
      return response.data.data
    },
    onSuccess: (_, draftId) => {
      queryClient.invalidateQueries({ queryKey: ['campaign-drafts', tenantId] })
      queryClient.invalidateQueries({ queryKey: ['campaign-draft', tenantId, draftId] })
    },
  })
}

export function useApproveDraft(tenantId: number) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: async (draftId: string) => {
      const response = await apiClient.post<{ data: CampaignDraft }>(
        `/campaign-builder/tenant/${tenantId}/campaign-drafts/${draftId}/approve`
      )
      return response.data.data
    },
    onSuccess: (_, draftId) => {
      queryClient.invalidateQueries({ queryKey: ['campaign-drafts', tenantId] })
      queryClient.invalidateQueries({ queryKey: ['campaign-draft', tenantId, draftId] })
    },
  })
}

export function useRejectDraft(tenantId: number) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: async ({ draftId, reason }: { draftId: string; reason: string }) => {
      const response = await apiClient.post<{ data: CampaignDraft }>(
        `/campaign-builder/tenant/${tenantId}/campaign-drafts/${draftId}/reject`,
        null,
        { params: { reason } }
      )
      return response.data.data
    },
    onSuccess: (_, { draftId }) => {
      queryClient.invalidateQueries({ queryKey: ['campaign-drafts', tenantId] })
      queryClient.invalidateQueries({ queryKey: ['campaign-draft', tenantId, draftId] })
    },
  })
}

export function usePublishDraft(tenantId: number) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: async (draftId: string) => {
      const response = await apiClient.post<{ data: CampaignDraft }>(
        `/campaign-builder/tenant/${tenantId}/campaign-drafts/${draftId}/publish`
      )
      return response.data.data
    },
    onSuccess: (_, draftId) => {
      queryClient.invalidateQueries({ queryKey: ['campaign-drafts', tenantId] })
      queryClient.invalidateQueries({ queryKey: ['campaign-draft', tenantId, draftId] })
      queryClient.invalidateQueries({ queryKey: ['publish-logs', tenantId] })
    },
  })
}

export function useDeleteCampaignDraft(tenantId: number) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: async (draftId: string) => {
      const response = await apiClient.delete(
        `/campaign-builder/tenant/${tenantId}/campaign-drafts/${draftId}`
      )
      return response.data
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['campaign-drafts', tenantId] })
    },
  })
}

// =============================================================================
// Publish Log Hooks
// =============================================================================

export function usePublishLogs(
  tenantId: number,
  filters?: {
    draft_id?: string
    platform?: Platform
    result_status?: PublishResult
  }
) {
  return useQuery({
    queryKey: ['publish-logs', tenantId, filters],
    queryFn: async () => {
      const response = await apiClient.get<{ data: PublishLog[] }>(
        `/campaign-builder/tenant/${tenantId}/campaign-publish-logs`,
        { params: filters }
      )
      return response.data.data
    },
    enabled: !!tenantId,
  })
}

export function useRetryPublish(tenantId: number) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: async (logId: string) => {
      const response = await apiClient.post(`/campaign-builder/tenant/${tenantId}/campaign-publish-logs/${logId}/retry`)
      return response.data
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['publish-logs', tenantId] })
      queryClient.invalidateQueries({ queryKey: ['campaign-drafts', tenantId] })
    },
  })
}

File: backend/app/models/campaign_builder.py
# =============================================================================
# Stratum AI - Campaign Builder Models
# =============================================================================
"""
Database models for the Campaign Builder feature:
- TenantPlatformConnection: OAuth tokens and connection metadata (per platform)
- TenantAdAccount: Ad accounts enabled for use
- CampaignDraft: Campaign drafts with approval workflow
- CampaignPublishLog: Audit trail for publish attempts
"""

import enum
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
from uuid import uuid4

from sqlalchemy import (
    Boolean,
    Column,
    DateTime,
)
from sqlalchemy import Enum as SQLEnum
from sqlalchemy import (
    ForeignKey,
    Index,
    Integer,
    Numeric,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import relationship

from app.db.base_class import Base

# =============================================================================
# Enums
# =============================================================================


class AdPlatform(str, enum.Enum):
    """Supported advertising platforms."""

    META = "meta"
    GOOGLE = "google"
    TIKTOK = "tiktok"
    SNAPCHAT = "snapchat"


class ConnectionStatus(str, enum.Enum):
    """Platform connection status."""

    CONNECTED = "connected"
    EXPIRED = "expired"
    ERROR = "error"
    DISCONNECTED = "disconnected"


class DraftStatus(str, enum.Enum):
    """Campaign draft status."""

    DRAFT = "draft"
    SUBMITTED = "submitted"
    APPROVED = "approved"
    REJECTED = "rejected"
    PUBLISHING = "publishing"
    PUBLISHED = "published"
    FAILED = "failed"


class PublishResult(str, enum.Enum):
    """Publish attempt result."""

    SUCCESS = "success"
    FAILURE = "failure"


# =============================================================================
# Models
# =============================================================================


class TenantPlatformConnection(Base):
    """
    Stores OAuth tokens and connection metadata.
    One record per platform.
    """

    __tablename__ = "tenant_platform_connection"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    platform = Column(String(50), nullable=False)
    status = Column(
        String(50), nullable=False, default=ConnectionStatus.DISCONNECTED.value
    )

    # Token storage (encrypted in production)
    token_ref = Column(
        Text, nullable=True
    )  # Reference to encrypted token in secrets manager
    access_token_encrypted = Column(Text, nullable=True)
    refresh_token_encrypted = Column(Text, nullable=True)
    token_expires_at = Column(DateTime(timezone=True), nullable=True)

    # OAuth metadata
    scopes = Column(JSONB, nullable=True, default=list)
    granted_by_user_id = Column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )

    # Timestamps
    connected_at = Column(DateTime(timezone=True), nullable=True)
    last_refreshed_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    # Error tracking
    last_error = Column(Text, nullable=True)
    error_count = Column(Integer, default=0)

    # Relationships
    granted_by = relationship("User", foreign_keys=[granted_by_user_id])
    ad_accounts = relationship(
        "TenantAdAccount", back_populates="connection", cascade="all, delete-orphan"
    )

    __table_args__ = (
        # was tenant-scoped; now global
        UniqueConstraint("platform", name="uq_platform_connection"),
    )


class TenantAdAccount(Base):
    """
    Ad accounts enabled for use in Stratum AI.
    Synced from platform after OAuth authorization.
    """

    __tablename__ = "tenant_ad_account"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    connection_id = Column(
        UUID(as_uuid=True),
        ForeignKey("tenant_platform_connection.id", ondelete="CASCADE"),
        nullable=False,
    )
    platform = Column(String(50), nullable=False)

    # Platform identifiers
    platform_account_id = Column(String(255), nullable=False)  # e.g., act_123456789
    name = Column(String(255), nullable=False)
    business_name = Column(String(255), nullable=True)

    # Account configuration
    currency = Column(String(10), nullable=False, default="USD")
    timezone = Column(String(100), nullable=False, default="UTC")
    is_enabled = Column(Boolean, default=False, nullable=False)

    # Budget controls
    daily_budget_cap = Column(Numeric(precision=12, scale=2), nullable=True)
    monthly_budget_cap = Column(Numeric(precision=12, scale=2), nullable=True)

    # Platform permissions and metadata
    permissions_json = Column(JSONB, nullable=True, default=dict)
    account_status = Column(String(50), nullable=True)  # active, disabled, etc.

    # Sync tracking
    last_synced_at = Column(DateTime(timezone=True), nullable=True)
    sync_error = Column(Text, nullable=True)

    # Timestamps
    created_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    # Relationships
    connection = relationship("TenantPlatformConnection", back_populates="ad_accounts")
    campaign_drafts = relationship("CampaignDraft", back_populates="ad_account")

    __table_args__ = (
        # was tenant-scoped; now global
        UniqueConstraint(
            "platform", "platform_account_id", name="uq_ad_account"
        ),
        Index("ix_ad_account_enabled", "is_enabled"),
    )


class CampaignDraft(Base):
    """
    Stores campaign drafts with approval workflow.
    Draft JSON is platform-agnostic, converted at publish time.
    """

    __tablename__ = "campaign_draft"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    ad_account_id = Column(
        UUID(as_uuid=True),
        ForeignKey("tenant_ad_account.id", ondelete="SET NULL"),
        nullable=True,
    )
    platform = Column(String(50), nullable=False)

    # Draft identification
    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    status = Column(String(50), nullable=False, default=DraftStatus.DRAFT.value)

    # Campaign configuration (canonical JSON format)
    draft_json = Column(JSONB, nullable=False, default=dict)

    # Workflow tracking
    created_by_user_id = Column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    submitted_by_user_id = Column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    approved_by_user_id = Column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    rejected_by_user_id = Column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )

    submitted_at = Column(DateTime(timezone=True), nullable=True)
    approved_at = Column(DateTime(timezone=True), nullable=True)
    rejected_at = Column(DateTime(timezone=True), nullable=True)
    rejection_reason = Column(Text, nullable=True)

    # Published campaign reference
    platform_campaign_id = Column(String(255), nullable=True)
    published_at = Column(DateTime(timezone=True), nullable=True)

    # Timestamps
    created_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    # Relationships
    ad_account = relationship("TenantAdAccount", back_populates="campaign_drafts")
    created_by = relationship("User", foreign_keys=[created_by_user_id])
    submitted_by = relationship("User", foreign_keys=[submitted_by_user_id])
    approved_by = relationship("User", foreign_keys=[approved_by_user_id])
    rejected_by = relationship("User", foreign_keys=[rejected_by_user_id])
    publish_logs = relationship(
        "CampaignPublishLog", back_populates="draft", cascade="all, delete-orphan"
    )

    __table_args__ = (
        Index("ix_campaign_draft_status", "status"),
        Index("ix_campaign_draft_platform", "platform"),
    )


class CampaignPublishLog(Base):
    """
    Audit trail for campaign publish attempts.
    Records request/response for debugging and compliance.
    """

    __tablename__ = "campaign_publish_log"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    draft_id = Column(
        UUID(as_uuid=True),
        ForeignKey("campaign_draft.id", ondelete="SET NULL"),
        nullable=True,
    )
    platform = Column(String(50), nullable=False)
    platform_account_id = Column(String(255), nullable=False)

    # Actor
    published_by_user_id = Column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )

    # Event timing
    event_time = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    # Request/Response (for debugging)
    request_json = Column(JSONB, nullable=True)
    response_json = Column(JSONB, nullable=True)

    # Result
    result_status = Column(String(50), nullable=False)
    platform_campaign_id = Column(String(255), nullable=True)  # If successful

    # Error details
    error_code = Column(String(100), nullable=True)
    error_message = Column(Text, nullable=True)

    # Retry tracking
    retry_count = Column(Integer, default=0)
    last_retry_at = Column(DateTime(timezone=True), nullable=True)

    # Relationships
    draft = relationship("CampaignDraft", back_populates="publish_logs")
    published_by = relationship("User", foreign_keys=[published_by_user_id])

    __table_args__ = (
        Index("ix_campaign_publish_log_time", "event_time"),
        Index("ix_campaign_publish_log_draft", "draft_id"),
    )
